### PR TITLE
useMultipleData

### PR DESCRIPTION
**Pull request by denis-sokolov**
This allows to use a single hook to retrieve or update multiple Oncilla atoms. See the usage in README.

This takes what we had as useData and modifies all individual work pieces to iterate over multiple data atoms.

### CODE REVIEW

**Comment by tjklemz** [comment]
Can you explain the use case? It's non-obvious to me what the difference between these is:
```javascript
useData("thing1", "1");
useData("thing2", "2");
```
versus
```javascript
useMultipleData({ thing1, thing2 })
```

**Comment by tjklemz** [comment]
Also, 19 days ago?!? Yikes! Our bad.

File: README.md [hunk 1]
```diff
@@ -33,6 +33,11 @@ return (
     />
   </div>
 );
+
+// Prefer multiple useData calls above, but when the data you need to fetch is dynamic in its length, useMultipleData can help:
+const [data, update] = useMultipleData({ tasks: ["1", "3"], jobs: ["2"] });
```
Comment (abelino) [inline README.md]:
```suggestion
const [data, update] = useMultipleData({ tasks: ["1", "3"], jobs: ["2"] });
// where `tasks` and `jobs` are atom kinds and strings in the array values are atom ids
```

Or something like that. I think we should explain the interface a little to the use in the readme.

Comment (tjklemz) [reply README.md]: I mean, I thought the interface was obvious, but there's two separate things going on here:
1) Multiple data kinds (`tasks` and `jobs` here).
2) Multiple ids in those kinds (the arrays).

I don't understand why these are grouped. IOW, the syncing ramifications are non-obvious to me. Are these all being batched? But doesn't `useData` already batch?

For multiple IDs, it seems to me that this would be better:
```javascript
useData("tasks", ["1", "3"])
useData("jobs", ["2"])
useData("other", "1")
```
where `useData` takes a scalar or an array for the IDs. And then they'd be accessed like so:
```
const [data] = useData("tasks", ["1", "3"])
const [task1, task3] = data;
```

Comment (tjklemz) [reply README.md]: Mostly, I think it'd be better to be reusing or reworking the existing `useData`. But I'd like to understand the problem a bit better, first.

Comment (denis-sokolov) [reply README.md]: We have this outstanding task to improve documentation, so you’re right. At the same time, you did understand the design, didn’t you? :-)

**Comment by denis-sokolov** [comment]
The reason for `useMultipleData` is you can have multiple ids you want to retrieve and using multiple `useData` hooks might be either too wordy in your source file, or even impossible if the number of ids is dynamic, since you can’t run React hooks in a loop.

While the simple use case of `const [task] = useData("tasks", "1")` remains the primary use to retrieve the task with id 1, you can now retrieve a lot of data in one go with the advanced, rarely needed `const [data] = useMultipleData({ tasks: taskIds, jobs: jobIds })`, which grabs multiple tasks and multiple jobs.

@tjklemz

File: README.md (hunk 1 above)
Comment (denis-sokolov) [reply README.md]: I think the suggestion about a hook that takes single kind, multiple ids is great. @tjklemz, thanks!

I don’t want to overload the simple `useData`, so I have instead overloaded the `useMultipleData`. This ensures that the simple use case is simple, and the advanced use cases are possible.

I have also slightly reshaped the return type. I think having the return `tasks` array be identified by the index in the array leads to awkward work aligning the id and the array index, so I instead return a map.

Please see the updated docs and test!

**Comment by tjklemz** [comment]
I like the improvements. I'd still argue for having one function, `useData`, like so:
```javascript
useData("tasks", "1")
useData("tasks", ["1", "3"])
useData({ tasks: ["1", "3"], jobs: ["1", "2"] })
```
I think it's less cognitive overhead to have one function. So if you could implement like the example I listed, I'd be ecstatic. Can you explain why you prefer having two different functions?

**Comment by denis-sokolov** [comment]
When I am expertly aware of some functionality, I love heavy overloads. One of my older favorite cute designs is a single `regex` function that simplifies work with regexes:
- `regex(/@/, str)` returns a boolean for conditionals (isEmailValid);
- `regex(/@(.+)$/, str)` returns a string or throws, for simple extractions (domain);
- `regex(/^(.+)@(.+)$/, str)` returns an array of strings or throws for multiple extractions (username, domain);
The general pattern being I have a powerful utility and I can call it in various ways and get exactly what I need, yeay.

However, when I’m not expertly aware of it, this causes a lot of very real problems:
- When I plan to use it one way, but make a mistake and pass wrong parameters, instead of getting an error, I get something completely different and unexpected;
- When I plan to use it one way, and suddenly see it used elsewhere in the codebase used another way, I get confused whether I’m doing it right;
- When I hover the function in the IDE, instead of getting a simple typehint popup directly showing how to use it, I get a garbage dump of overloads of complex TypeScript types;
- When I search for the function in the documentation I get many more results than is helpful;

I follow a very important design guide for me – to make simple things simple, and complex things possible. With the proposed design, day-to-day developers using the simple useData have one way to use it, can’t screw it up, see it everywhere in the code the same way, and are happy to have it straight-forward. For those rare cases where me or you need a complex case, we can reach for the “big guns” and grab useMultipleData.

Did I convince you that it’s not an awful idea to have the two separated? :-)

**Comment by tjklemz** [comment]
@abelino @nateq314 @amclain any thoughts on the above convo?

@denis-sokolov I'm still not convinced, since the doc/help argument doesn't simplify. I have to know about the different functions and what they do. :) So I'm still arguing for one function, `useData`. But I'll leave it to a vote, since I don't think it's an absolutely "awful idea to have the two separated" either.

**Comment by abelino** [comment]
@denis-sokolov @tjklemz I am in favor of not overloading `useData` a whole lot... at most:
```
useData("tasks", "1")
useData("tasks", ["1", "3"])
```

`useData`, just by reading its signature identifier, tells me that it'll only fetch a single `kind`... meaning data that has compatible structures. And having the ability to pull multiple `id`s through `useData` is convenient and also doesn't completely bloat the api.

I also agree agree with how `useMultipleData`... again from the name alone I can wrap my head around the api returning different `kind`s which could possible have different structures.

**Comment by denis-sokolov** [comment]
@tjklemz, the point is you _don’t_ have to know about different functions! Most of your teammates will only ever use `useData`. `useMultipleData` will be hidden deeper in the docs, because it’s a tool only a few devs will reach out in a few select cases.

**Comment by amclain** [comment]
Is the underlying debate that's going on at this point around the naming of the functions? For example, if we had functions `useKind` and `useKinds`, does that conceptually find the right balance between having separate functions, while also maintaining an intuitive API? I don't care what the actual function names are, but just pointing out that it sounds like the contention is around `useMultipleData` being slightly too distant from `useData` which makes it tempting to overload the `useData` interface, whereas in my example it's more natural to understand that "I add an 's' to make the word plural and therefore I'm able to use a slightly different interface".

**Comment by tjklemz** [comment]
@denis-sokolov let me talk this out and summarize the discussion. Please bear with me.

You're claiming that `useMultipleData` will be a non-standard use case. But here is something that comes up all the time in our codebase:
```javascript
const [tasks] = useData("tasks", "1")
const [jobs] = useData("jobs", "1")
const [stuff] = useData("stuff", "1")
if (tasks === "loading" || jobs === "loading" || stuff === "loading") return null;
```
Super common. I want "tasks" and "jobs" and "stuff". And I can't do anything until I have all of it. This is why I like the object format, actually (if I'm understanding your code correctly):
```javascript
const [data] = useData({ "tasks": ["1", "3"], "jobs": ["1", "2"], "stuff": ["1"] })
if (data === "loading") return null;
// Alternatively, you could have the loading state on each key. I would do this interface if I were _always_ requiring an object be passed to useData.
if (Object.values(data).some(v => v === "loading")) return null;
```
_But_, if I have that use case, I should probably create a separate/custom hook, right?. After all, _why_ are they related?

So. That leaves us with multiple IDs.

IOW, drop the object format, like you were saying in Slack:
```javascript
const [tasks] = useData("tasks", "1")
const [jobs] = useData("jobs", ["1", "2"])
```
Your claim here, though, is that this should be:
```javascript
const [tasks] = useData("tasks", "1")
const [jobs] = useMultipleData("jobs", ["1", "2"])
```
And I can live with that if there's no object form. There's two options for the IDs here, and it would get annoying if the interface were identical, from a TypeScript standpoint (i.e. on hover). Plus, semantically, the result from the backend should be a list or a scalar, not one or the other. You could have it be jQuery-esque, where you always require an array, but I think that complicates the usage here in React-land.

So. *tl;dr*—Can we agree on that last example? `useData` takes a scalar for an ID. `useMultipleData` takes an array of IDs.

**Comment by denis-sokolov** [comment]
@tjklemz, as discussed, updated, take another look! There is still the map, but it’s only used internally.
